dds/idl/field_info.cpp

```diff
@@ -0,0 +1,168 @@
+{
+  return "IDL::DistinctType<" + const_s + scoped_t + ", " + underscored_t + "_tag>";
+}
+
+// for anonymous types
+FieldInfo::FieldInfo(AST_Field& field) :
+  type_(field.field_type()),
+  name_(field.local_name()->get_string()),
+  underscored_(underscored_type_name(type_->name(), *type_))
+{
+  init();
+}
+
+FieldInfo::FieldInfo(UTL_ScopedName* sn, AST_Type* base) :
+  type_(base),
+  underscored_(underscored_type_name(sn, *type_)),
+  scoped_type_(scoped(sn))
+{
+  init();
+}
+
+void FieldInfo::init()
+{
+  act_ = resolveActualType(type_);
+  cls_ = classify(act_);
+  arr_ = AST_Array::narrow_from_decl(type_);
+  seq_ = AST_Sequence::narrow_from_decl(type_);
+  as_base_ = arr_ ? arr_->base_type() : (seq_ ? seq_->base_type() : 0);
+  as_act_ = as_base_ ? resolveActualType(as_base_) : 0;
+  as_cls_ = as_act_ ? classify(as_act_) : CL_UNKNOWN;
+  scoped_elem_ = as_base_ ? scoped(as_base_->name()) : "";
+  underscored_elem_ = as_base_ ? underscore(scoped_elem_) : "";
+  if (!scoped_elem_.empty() && !underscored_elem_.empty()) {
+    elem_ref_ = ref(scoped_elem_, underscored_elem_, "");
+    elem_const_ref_ = ref(scoped_elem_, underscored_elem_);
+  }
+
+  if (type_->anonymous() && as_base_) {
+    scoped_type_ = scoped(type_->name());
+    std::size_t i = scoped_type_.rfind(scope_op);
+    struct_name_ = scoped_type_.substr(0, i);
+    if (!name_.empty()) {
+      type_name_ = at_pfx() + name_;
+      if (seq_) { type_name_ += "_seq"; }
+      scoped_type_ = struct_name_ + "::" + type_name_;
+      underscored_ = underscore(scoped_type_);
+    } else {
+      type_name_ = scoped_type_.substr(i + 2);
+    }
+  } else {
+    if (scoped_type_.empty()) {
+      scoped_type_ = scoped(type_->name());
+    }
+    //name_
+    type_name_ = scoped_type_;
+    //struct_name_
+  }
+
+  n_elems_ = 1;
+  if (arr_) {
+    for (size_t i = 0; i < arr_->n_dims(); ++i) {
+      n_elems_ *= arr_->dims()[i]->ev()->u.ulval;
+    }
+    std::ostringstream os;
+    os << n_elems_;
+    length_ = os.str();
+    arg_ = "arr";
+  } else if (seq_) {
+    n_elems_ = !seq_->unbounded() ? seq_->max_size()->ev()->u.ulval : 0;
+    length_ = "length";
+    arg_ = "seq";
+  }
+
+  if (cxx11()) {
+    unwrap_ = scoped_type_ + "& " + arg_ + " = wrap;\n  ACE_UNUSED_ARG(" + arg_ + ");\n";
+    const_unwrap_ = "  const " + unwrap_;
+    unwrap_ = "  " + unwrap_;
+    arg_ = "wrap";
+    ref_ = ref(scoped_type_, underscored_, "");
+    const_ref_ = ref(scoped_type_, underscored_);
+    ptr_ = "IDL::DistinctType<" + scoped_type_ + ", " + underscored_ + "_tag>*";
+  } else {
+    ref_ = scoped_type_ + (arr_ ? "_forany&" : "&");
+    const_ref_ = "const " + ref_;
+    ptr_ = scoped_type_ + (arr_ ? "_forany*" : "*");
+  }
+}
+
+bool FieldInfo::is_new(EleLenSet& el_set)
+{
+  return cxx11() || el_set.insert(EleLen(*this)).second;
+}
+
+bool FieldInfo::anonymous_array() const
+{
+  return type_->anonymous() && arr_ && (cls_ & CL_ARRAY);
```

Comment:
I would think arr_ is sufficient. During the development, these 3 methods anonymous_array(), anonymous_sequence(), and anonymous() were factored out from different places, where some checked the non-null pointer and some checked classification constant also. The AST_Type::anonymous() was seen giving the incorrect values in some place. In order to fix those problems, I replaced most of such checks with one of these 3 (probably over-kill) checking methods. Since this is the code-generation code, not the runtime application code, an extra comparison won't hurt much.

Reply:
Removed anonymous_array() and anonymous_sequence().